File: 🟡 Medium/1438 - Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/_1438LongContSubArrWithAbsDiffLessThanOrEqualToLimit..cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        deque<int> inc;
        deque<int> dec;
        int l = 0;
        int res = 0;

        for (int r = 0; r < nums.size(); r++) {
            while (!inc.empty() && inc.back() > nums[r]) {
                inc.pop_back();
            }

            inc.push_back(nums[r]);

            while (!dec.empty() && dec.back() < nums[r]) {
                dec.pop_back();
            }

            dec.push_back(nums[r]);

            while (dec.front() - inc.front() > limit) {
                if (dec.front() == nums[l]) {
                    dec.pop_front();
                }
                if (inc.front() == nums[l]) {
                    inc.pop_front();
                }
                l++;
            }
            res = max(res, r - l + 1);
        }
        return res;
    }
};

File: 🟡 Medium/1438 - Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/_1438LongContSubArrWithAbsDiffLessThanOrEqualToLimit.java
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        Deque<Integer> inc = new LinkedList<>();
        Deque<Integer> dec = new LinkedList<>();
        int l = 0;
        int res = 0;

        for (int r = 0; r < nums.length; r++) {
            while (!inc.isEmpty() && inc.getLast() > nums[r]) {
                inc.removeLast();
            }
            while (!dec.isEmpty() && dec.getLast() < nums[r]) {
                dec.removeLast();
            }

            inc.addLast(nums[r]);
            dec.addLast(nums[r]);

            while (dec.getFirst() - inc.getFirst() > limit) {
                if (dec.getFirst() == nums[l]) {
                    dec.removeFirst();
                }
                if (inc.getFirst() == nums[l]) {
                    inc.removeFirst();
                }
                l++;
            }
            res = Math.max(res, r - l + 1);
        }
        return res;
    }
}